[Feat] 관리자 로그인/로그아웃 기능 구현

CUCU7103 opened the pull request:
## History

이슈번호: #482 

## 🚀 Major Changes & Explanations

<!--빠른 리뷰를 위해 이해를 도울 만한 설명을 자세히 적어주세요-->
- 관리자 로그인, 로그아웃 기능을 구현
- id ,pw 방식의 전통적인 로그인 방식 + jwt
- member 테이블 role 컬럼 제거
- member 엔티티  클래스 UserDetails 상속 관계 제거
- 패키지 수정 (token => auth)
  - auth 하위에 로그인/로그아웃 관련 코드를 admin / customer / seller 나눠서 관리 
- principal 타입 수정(Long => BbangleUserPrincipal)
- refresh_token 테이블 수정
  - 컬럼명 수정 member_id  => user_id
  - 컬럼 추가 user_role
  - user_id + user_role 유니크 제약조건 추가

## 📷 Test Image

<!-- postman, swagger 등을 활용한 api 결과, 각종 Edge case 테스트 결과 이미지를 붙여주세요-->
<!-- 이미지가 많거나 클 경우 오른쪽 패턴을 이용해주세요<img src = "CREATED_IMG_URL" width = "400px">-->
<img width="2381" height="298" alt="image" src="https://github.com/user-attachments/assets/794adf07-55a6-4123-96f5-b1fe4bd3419d" />

<img width="365" height="159" alt="image" src="https://github.com/user-attachments/assets/fa028d8a-7320-4c38-aa27-ca9a26d70bbe" />


## 💡 ETC

<!-- ex) 질문. 작업 관련 사항, 고민한 내용 등등을 적어주세요-->
- 관리자 가입방식은 SQL문으로 직접 DB insert
  - pw의 경우 https://bcrypt-generator.com 를 통해서 인코딩 후 그 값을 컬럼에 저장

CUCU7103 committed:
[Feat] 관리자 도메인 생성
1. 관리자 도메인 생성
2. 관리자 테이블 생성

CUCU7103 committed:
[Feat] RefreshToken 수정
1. 관리자 로그인을 위한 도메인 수정
2. Repository 수정
3. 테이블 컬럼 추가

CUCU7103 committed:
[Feat] 관리자 로그인/로그아웃 기능 추가
1. 관리자 로그인 기능 구현
2. Security 설정 수정
3. 관리자 로그아웃 기능 추가
4. 패키지 구조 변경으로 인한 수정

CUCU7103 committed:
[Refactor] 로그인 로직 , flyway 파일 수정
1. 로그인 시 passwordEncoder 추가
2. flyway sql문 추가

CUCU7103 committed:
[Refactor] 테스트 개선 및 추가
1. 통합 테스트 추가
2. 단위 테스트 개선
3. 테스트 설정 추가

CUCU7103 committed:
[Refactor] 로그아웃 기능 수정 및 테스트 코드 추가
1. 로그아웃 기능 개선
2. 로그아웃 테스트 코드 추가

kmindev commented on src/main/java/com/bbangle/bbangle/admin/admin/controller/AdminAuthController.java:
```diff
@@ -0,0 +1,37 @@
+package com.bbangle.bbangle.admin.admin.controller;
```
로그인 관련 api는 token 패키지 내에 위치하고 있는데, admin보다는 token 패키지로 가는게 어떠신가요?

CUCU7103 replied:
@kmindev 
제 생각으로는 현재 토큰 도메인에 로그인 기능이 존재하는게 조금 맞지 않는다고 생각되어집니다
차라리 auth나 login과 같이 좀 더 인증/인가라는 큰 범위의 도메인 하위에 토큰이 들어가야 된다고 생각해서 일단 분리시켜놓았습니다

kmindev replied:
token이 상위 패키지에 있어서 의미가 좀 불분명하긴 한 것 같네요...

그럼 auth 패키지를 하나 만들어서
```
auth
- customer
- seller
- admin
```

이렇게 나누는 거 어떠신가요?

kmindev commented on src/main/java/com/bbangle/bbangle/admin/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,37 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/admin")
+public class AdminController implements AdminApi {
```
로그인·로그아웃처럼 인증 역할만 수행하는 컨트롤러라면 AdminController보다는 AdminAuthController가 책임을 더 명확하게 드러낼 것 같습니다. 일반적으로 Admin은 관리자 기능 전체를 의미하기 때문에, 인증 영역을 구분하기 위해 네이밍을 분리하는 것이 도움이 될 것 같습니다!

url도 /api/v1/admin/auth/login 이런식으로 가는것 도 괜찮을 것 같네요

CUCU7103 replied:
@kmindev  
좋은 생각입니다 인증 역할만 수행하는 컨트롤러는 말씀하신대로 네이밍을 변경하는게 적합하다고 생각되어집니다
해당 부분은 수정 진행하겠습니다.

kmindev commented on src/main/java/com/bbangle/bbangle/admin/admin/service/AdminService.java:
```diff
@@ -0,0 +1,71 @@
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class AdminService {
```
여기도 마찬가지로 네이밍을 AdminAuthService 이렇게 가져가는게 좋을 것 같습니다.

CUCU7103 replied:
@kmindev 
네네 수정 진행하겠습니다~

kmindev commented on src/main/java/com/bbangle/bbangle/auth/domain/RefreshToken.java:
```diff
@@ -25,9 +25,12 @@ public class RefreshToken {
+    @Column(name = "member_id", unique = true)
     private Long memberId;
 
+    @Column(name = "admin_id", unique = true)
```
adminId 컬럼을 추가하게 되면 이후 셀러 등 다른 주체가 추가될 때마다 동일한 방식으로 컬럼이 계속 증가하게 되고, 자연스럽게 null 컬럼도 늘어나는 구조가 될 것으로 보입니다. 이러한 형태는 확장성과 정규화 측면에서 개선 여지가 있어보입니다.

대안으로 3가지 방법 정도 생각나는데, 
1. RefreshToken 테이블을 사용자 유형별로 분리
2. id(member_id, seller_id, admin_id) + role을 해서 유니키 키하게 구성하는 방법
3. refreshToken 내에 이미 member_id을 포함하고 있기에 member_id를 제거하는 방법도 있는데, 이 방법은 특정 사용자의 refresh token을 조회하거나 무효화하는 과정이 오히려 어려워질 가능성이 있어보이네요

CUCU7103 replied:
@kmindev 
refresh token의 경우 역할은 인증을 위한 토큰 저장소의 역할 이기에  크게 변화가 있을 것 같지는 않습니다
차후 확장성을 고려한다면 아예 redis로 관리하는게 확실하다고 생각되어집니다, (세션 저장소)
현 상황에서는 2번 방법이 가장 효과적일 것 같습니다!

kmindev replied:
@CUCU7103 준규님 이 부분도 애기가 필요했는데 깜빡했네요..
이 부분도 2번 방식으로 우선 진행하겠습니다

kmindev commented on src/main/java/com/bbangle/bbangle/auth/admin/service/AdminAuthService.java:
```diff
@@ -0,0 +1,71 @@
+        refreshToken.update(refreshTokenVal);
+        refreshTokenRepository.save(refreshToken);
+
+        return AdminLoginResponse.builder()
```
저는 DTO나 Entity 객체 생성할 때 빌더 패턴보다 생성자 팩토리 메서드 패턴을 추구하는데, 이것도 명확한 기준이 필요할 것 같네요..

저는 빌더 패턴을 사용하면 코드 길이가 길어지고, 호출하는 쪽에서 알아야할 정보들이 많아서 별로 추구하지 않습니다. 
생성자 팩토리 메서드를 통해 호출하는 쪽에서 필요한 데이터가 무엇인지 확실하게 제공하기에 이점이 있다고 생각합니다. 

준규님은 빌더 패턴을 선호하시는 이유가 있으신가요?

kmindev commented on src/main/java/com/bbangle/bbangle/common/role/Role.java:
```diff
@@ -8,7 +8,7 @@ public enum Role {
     ROLE_USER("ROLE_USER", "사용자"),
     ROLE_SELLER("ROLE_SELLER", "판매자"),
-    ROLE_ADMIN("ROLE_ADMIN", "빵그리오븐관리자");
+    ROLE_ADMIN("ROLE_ADMIN", "빵그리오븐관리자"); // prefix 제거
```
주석이 의미하는 바가 어떤걸까요!!??

CUCU7103 replied:
@kmindev 
아 저 주석은 오타입니다 ㅋㅋㅋㅋ;;; 수정하겠습니다

kmindev commented on src/main/java/com/bbangle/bbangle/config/WebOAuthSecurityConfig.java:
```diff
@@ -61,6 +62,8 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                 UsernamePasswordAuthenticationFilter.class)
             .authorizeHttpRequests(authorize ->
                 authorize.requestMatchers("/api/v1/token").permitAll()
+                    .requestMatchers("/api/v1/admin/login").permitAll()
```
권한 깔끔하게 나누는 것도 고민을 한 번 해봤는데
- public api (모두 가능)
- 고객 api
- 판매자 api
- 관리자 api 

4가지 경우로 정의하면 좋을 것 같습니다
이 4가지를 클래스를 만들어서 관리하는게 유지보수면에서 유리할 것 같은데 제가 리팩토링 진행해봐도 될까요?

CUCU7103 replied:
@kmindev 
네네 좋습니다 리펙터링 진행하시죠!!

kmindev commented on src/main/resources/flyway/V25__app.sql:
```diff
@@ -0,0 +1,9 @@
+-- Add admin_id column to refresh_token table
```
flyway 파일 두 개로 나누신 이유가 있으신가요?

CUCU7103 replied:
@kmindev 
해당 부분은 작업하다보니 V24를 커밋한 뒤에 다른 테이블 수정사항이 생겨서 두번에 걸쳐서 올렸습니다
합치는게 좋을까요???

CUCU7103 committed:
[Refactor] 오타 및 클래스 네이밍 변경

kmindev committed:
[Refactor] API Path 클래스 도입
- SecuirtyConfig 에서 api 권한을 위해 모든 api 정보를 담고 있어서 가독성과 유지보수에 어려움이 있었음. 권한별 클래스를 나누어서 관리하도록 리팩토링함.

kmindev committed:
Merge remote-tracking branch 'origin/feature/admin-login' into feature/admin-login

kmindev commented on src/main/java/com/bbangle/bbangle/admin/admin/controller/swagger/AdminApi.java:
```diff
@@ -0,0 +1,23 @@
+import io.swagger.v3.oas.annotations.tags.Tag;
+
+@Tag(name = "Admin", description = "관리자 API")
+public interface AdminApi {
```
이건 네이밍 수정이 안된 것 같네요!
제가 수정하겠습니다!

kmindev committed:
[Refactor] 클래스 네이밍 수정
- AdminApi => AdminAuthApi로 수정

kmindev committed:
[Feat] 테스트가 깨지는 문제를 해결
- /seller/** API 권한을 seller로 변경했더니 테스트가 깨지는 문제가 있었다.
- seller 권한을 부여하여 테스트 깨지는 문제 해결함.

kmindev commented on src/main/java/com/bbangle/bbangle/config/security/jwt/TokenProvider.java:
준규님! 
principal에 단순 Long id가 아닌 사용자 별로 구분할 수 있게 커스텀 객체 구성하는 방식은 고민해보셨나요?
지난번에 말씀 드렸듯이 단순 id만 넘겨주는 방식은 권한은 확인할 수 있지만, controller에서 어떤 사용자(고객, 셀러, 관리자) 인지 구분이 안되는 문제가 있습니다!
제가 생각했을 때 커스텀 객체를 구현하는 방안이 좋을 것 같습니다!
토요일에 한번 이야기 해보져~

```java
 public Authentication getAuthentication(String token) {
    return new UsernamePasswordAuthenticationToken(memberId, token, authorities);
}
```

토요일에 이야기 할 거 여기 간단하게 메모하겟습니다.
- Member 테이블에 role 컬럼도 더이상 필요가 없을 것 같음.
- Member 테이블 상속관계의 UserDetails 삭제해도 될 것 같음.
- Controller에서 principal 객체를 식별하기 위한 방법
- 일정 산정
위 내용 정리하고, 이번주말 내에 머지 진행하시죠!

kmindev committed:
[Refactor]: jwt 패키지 config/security 하위로 이동

kmindev committed:
[Refactor]: 패키지 네이밍 변경(token => auth)과 auth 패키지 분리 작업(admin, seller, customer) 기준으로 패키지 분리 작업
- admin, seller, customer 기준으로 패키지 분리 작업

kmindev committed:
[Feat]: member 테이블에 role 컬럼 제거, member 엔티티 클래스 userDetails 상속 관계 끊기
- member 테이블에 더이상 role 컬럼이 필요없어져서 삭제함.
- 불필요한 상속관계 제거

kmindev committed:
[Feat]: Principal 타입 변경 Long => 커스텀 Principal 객체
- 로그인 구조가 기존에는 member 테이블로만 처리를 했는데, seller, admin 테이블이 새로 추가됨에 따라 로그인 시 각각의 테이블 보고 로그인을 처리함 => 이는 로그인 시 Principal로 저장되는 id(pk)의 중복이 발생하고, 이는 컨트롤러에서 어떤 사용자 유형인지 구분할 수 가 없다.
- 그에 따라 커스텀 Principal 객체를 만들어 사용하기로 결정했다.

kmindev committed:
[Refactor]: flyway sql 파일 통합

kmindev committed:
[Feat]: RefreshToken 테이블 및 엔티티 클래스 수정
- member_id => user_id
- user_role 추가
- user_id + user_role 유니크 제약조건 추가

kmindev committed:
[Feat]: 에러메시지 수정

CUCU7103 commented on src/main/java/com/bbangle/bbangle/config/security/PublicApiPath.java:
@kmindev 
경민님 로그인 기능은 퍼블릭으로 두셨는데 로그아웃은 별도로 빼신 이유가 궁금합니다!!
보통 둘다 퍼블릭으로 두지 않나요??

kmindev committed:
[Feat]: 로그아웃 API public API로 변경
